core/src/ids.rs

```diff
@@ -205,11 +205,19 @@ impl ProgramId {
 
     /// Generate ProgramId from given CodeId and salt
     pub fn generate(code_id: CodeId, salt: &[u8]) -> Self {
-        const SALT: &[u8] = b"program";
+        const SALT: &[u8] = b"program_from_user";
 
         let argument = [SALT, code_id.as_ref(), salt].concat();
         hash(&argument).into()
     }
+
+    /// Generate ProgramId from given CodeId, MessageId and salt
+    pub fn generate_with_id(code_id: CodeId, salt: &[u8], message_id: MessageId) -> Self {
+        const SALT: &[u8] = b"program_from_wasm";
+
+        let argument = [message_id.as_ref(), SALT, code_id.as_ref(), salt].concat();
```

Comment:
```suggestion
        let argument = [SALT, message_id.as_ref(), code_id.as_ref(), salt].concat();
```

just for consistency